### ISSUE TITLE
Missing warning for migration of `let _ = x` between Rust 2018 and 2021 edition

### ISSUE THREAD

**Issue by sdroege** [rust-lang/rust#90465]
The following example  does not get a migration warning in Rust 2018 (or Rust 2021) -- ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8841e3a3e754a39fa51f448edb136af1)):

```rust
#![warn(rust_2021_incompatible_closure_captures)]

fn main() {
    struct Foo(u32);
    impl Drop for Foo {
        fn drop(&mut self) {
            println!("dropped {}", self.0);
        }
    }
    
    let f0 = Foo(0);
    let f1 = Foo(1);
    
    let c0 = move || {
        let _ = f0;
    };
    
    let c1 = move || {
        let _ = &f1;
    };
    
    println!("dropping 0");
    drop(c0);
    println!("dropping 1");
    drop(c1);
    println!("dropped all");
}
```

The behavior of `let _ = x` changed in the new edition, but there ought to have been a warning.

<details>
<summary>Original issue</summary>

See the below code

```rust
fn main() {
    struct Foo(u32);
    impl Drop for Foo {
        fn drop(&mut self) {
            println!("dropped {}", self.0);
        }
    }
    
    let f0 = Foo(0);
    let f1 = Foo(1);
    
    let c0 = move || {
        let _ = f0;
    };
    
    let c1 = move || {
        let _ = &f1;
    };
    
    println!("dropping 0");
    drop(c0);
    println!("dropping 1");
    drop(c1);
    println!("dropped all");
}
```

With 2018 edition ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e8b0d8c2e4d9709b40639bf4efa05a88)) this prints

```
dropping 0
dropped 0
dropping 1
dropped 1
dropped all
```

With 2021 edition ([playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=489fea42e35157e720ffef1d3fe6bffc)) this prints

```
dropping 0
dropping 1
dropped 1
dropped all
dropped 0
```

That is, with the 2021 edition the first closure is apparently not capturing `f0` at all (and drops it at the end of `main`) while with the 2018 edition it did.

This might be intentional and expected or not, which is why I'm creating this issue. It might be worth mentioning it as a potential porting trap in the edition guide. Note that `cargo fix --edition` did not catch this one (but a couple of others where the behaviour difference was irrelevant).

This subtle behaviour change caused [a test in gtk-rs](https://github.com/gtk-rs/gtk-rs-core/pull/321) to timeout and fail.

</details>

**Comment by ehuss** [comment]
Thanks for the report!  I believe this is already mentioned in the edition guide [here](https://doc.rust-lang.org/nightly/edition-guide/rust-2021/disjoint-capture-in-closures.html#wild-card-patterns).  Values that aren't read are not captured, and I believe that is the intended behavior.

However, I am a little surprised that `rust_2021_incompatible_closure_captures` does not add a suggestion to add `let _ = &f0;` to it.  @rust-lang/wg-rfc-2229 was that intended or a known hole in the migration?

**Comment by sdroege** [comment]
The confusing issue here is that the value was actually "used" (well, it was written in the closure body) and only the easy to miss `&` was missing.

Maybe there should be generally a lint for the `let _ = x` case in closure bodies as that's probably never what one intended?

**Comment by nikomatsakis** [comment]
This is indeed expected behavior, but a lint for this case is an interesting idea. I guess that the `let _ = x` was there previously in order to force capture, @sdroege?

**Comment by nikomatsakis** [comment]
Hmm, I too am a bit surprised that you don't get a migration lint already. Are you sure that you don't?

**Comment by nikomatsakis** [comment]
OK, I can attest that you don't get a warning ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=fdbcd42f10e48b9a7a8797bf2b687000)). Curious.

**Comment by nikomatsakis** [comment]
I agree that `let _ = x` should potentially issue a warning, but I don't think it's connected to closures per se. It also has no effect in regular functions. For example, this compiles, even though `x` is not initialized: ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e169b4e04a568a4c0efe92839b211ad3))

```rust
fn foo() {
    let x: u32;
    let _ = x; 
}
```

**Comment by nikomatsakis** [comment]
@sdroege +1 from me, I think @Mark-Simulacrum has often discussed their desire for such a lint as well

**Comment by sdroege** [comment]
Sure, see https://github.com/rust-lang/rust/issues/90524 . Thanks

**Comment by nikomatsakis** [comment]
Fix in #90597
